crash_rate_aggregates/fill_database.py

```diff
@@ -0,0 +1,192 @@
+        if not isinstance(activity_date, str) and not isinstance(activity_date, unicode):
+            return False
+        subsession_length = ping["payload/info/subsessionLength"]
+        if subsession_length is not None and not isinstance(subsession_length, int) and not isinstance(subsession_length, long):
+            return False
+        return True
+
+    def get_crash_pair(ping): # responsible for normalizing a single ping into a crash pair
+        # we need to parse and normalize the dates here rather than at the aggregates level,
+        # because we need to normalize and get rid of the time portion
+
+        # date the ping was received
+        submission_date = datetime.strptime(ping["meta/submissionDate"], "%Y%m%d").date() # convert the YYYYMMDD format to a real date
+        submission_date = max(start_date, min(end_date, submission_date)) # normalize the submission date if it's out of range
+
+        # date the ping was created on the client
+        try:
+            activity_date = dateutil.parser.parse(ping["creationDate"]).date() # the activity date is the date portion of creationDate
+        except: # malformed date, like how sometimes seconds that are not in [0, 59]
+            activity_date = submission_date # we'll just set it to the submission date, since we can't parse the activity date
+        activity_date = max(submission_date - timedelta(days=7), min(submission_date, activity_date)) # normalize the activity date if it's out of range
+
+        return (
+            # the keys we want to filter based on
+            (submission_date, activity_date) + tuple(ping[key] for key in comparable_dimensions), # all the dimensions we can compare by
+            # the crash values
+            np.array([
+                min(25, (ping["payload/info/subsessionLength"] or 0) / 3600.0), # usage hours, limited to ~25 hours to keep things normalized
+                int(ping["meta/docType"] == "crash"), # main crash (is a crash ping)
+                ping["payload/keyedHistograms/SUBPROCESS_CRASHES_WITH_DUMP/content_parent"] or 0, # content process crashes
+                (ping["payload/keyedHistograms/SUBPROCESS_CRASHES_WITH_DUMP/plugin_parent"] or 0), # plugin crashes
+                (ping["payload/keyedHistograms/SUBPROCESS_CRASHES_WITH_DUMP/gmplugin_parent"] or 0), # GMplugin crashes
+            ])
+        )
+    crash_values = ping_properties.filter(is_valid).map(get_crash_pair).reduceByKey(lambda a, b: a + b)
+
+    def dimension_mapping(pair): # responsible for converting aggregate crash pairs into individual dimension fields
+        dimension_key = pair[0]
+        (submission_date, activity_date), dimension_values = dimension_key[:2], dimension_key[2:]
+        usage_hours, main_crashes, content_crashes, plugin_crashes, gmplugin_crashes = pair[1]
+        return (
+            submission_date, activity_date,
+            {
+                key: str(dimension_value) if dimension_value is not None else ""
```

Comment:
We should come up with a better way to represent `null` values that doesn't use the empty string.